clash-prelude/src/Clash/Class/Counter/Internal.hs

```diff
@@ -0,0 +1,223 @@
+-- once all counters reach their maximum they reset to zero - i.e. odometer
+-- rollover. See 'Clash.Class.Counter.countSucc' and 'Clash.Class.Counter.countPred'
+-- for API usage examples.
+--
+-- Example use case: when driving a monitor through VGA you would like to keep
+-- track at least two counters: one counting a horizontal position, and one
+-- vertical. Perhaps a fancy VGA driver would also like to keep track of the
+-- number of drawn frames. To do so, the three counters are setup with different
+-- types. On each /round/ of the horizontal counter the vertical counter should
+-- be increased. On each /round/ of the vertical counter the frame counter should
+-- be increased. With this class you could simply use the type:
+--
+-- @
+-- (FrameCount, VerticalCount, HorizontalCount)
+-- @
+--
+-- and have 'Clash.Class.Counter.countSucc' work as described.
+--
+-- __N.B.__: This class exposes four functions 'countMin', 'countMax',
+-- 'countSuccOverflow', and 'countPredOverflow'. These functions are considered
+-- an internal API. Users are encouraged to use 'Clash.Class.Counter.countSucc'
+-- and 'Clash.Class.Counter.countPred'.
+--
+class Counter a where
+  -- | Value counter wraps around to on a 'countSuccOverflow' overflow
+  countMin :: a
+  default countMin :: Bounded a => a
+  countMin = minBound
+
+  -- | Value counter wraps around to on a 'countPredOverflow' overflow
+  countMax :: a
+  default countMax :: Bounded a => a
+  countMax = maxBound
+
+  -- | Gets the successor of @a@. If it overflows, the left part of the tuple
+  -- will be set to True.
+  countSuccOverflow :: a -> (Bool, a)
+  default countSuccOverflow :: (Eq a, Enum a, Bounded a) => a -> (Bool, a)
+  countSuccOverflow a
+    | a == maxBound = (True, countMin)
+    | otherwise = (False, succ a)
+
+  -- | Gets the predecessor of @a@. If it overflows, the left part of the tuple
+  -- will be set to True.
+  countPredOverflow :: a -> (Bool, a)
+  default countPredOverflow :: (Eq a, Enum a, Bounded a) => a -> (Bool, a)
+  countPredOverflow a
+    | a == minBound = (True, countMax)
+    | otherwise = (False, pred a)
+
+instance (1 <= n, KnownNat n) => Counter (Index n)
+instance KnownNat n => Counter (Unsigned n)
+instance KnownNat n => Counter (Signed n)
+instance KnownNat n => Counter (BitVector n)
+
+-- | Counter instance that flip-flops between 'Left' and 'Right'. Examples:
+--
+-- >>> type T = Either (Index 2) (Unsigned 2)
+-- >>> countSucc @T (Left 0)
+-- Left 1
+-- >>> countSucc @T (Left 1)
+-- Right 0
+-- >>> countSucc @T (Right 0)
+-- Right 1
+instance (Counter a, Counter b) => Counter (Either a b) where
+  countMin = Left countMin
+  countMax = Right countMax
+
+  countSuccOverflow e =
+    case bimap countSuccOverflow countSuccOverflow e of
+      Left (overflow, a)  -> (False, if overflow then Right countMin else Left a)
+      Right (overflow, b) -> (overflow, if overflow then Left countMin else Right b)
```

Comment:
Sadly, Clash's case-of-case transformation only fires when the subject of the outer case (the alternatives of the inner case) are non-representable: https://github.com/clash-lang/clash-compiler/blob/12fa89070d2cf55c133ab93af2faab2b611659f3/clash-lib/src/Clash/Normalize/Transformations.hs#L392-L397

I don't know whether GHC does a case-of-case transformation, but if it doesn't then this code introduces two multiplexers where one would be sufficient. You should check the output. If a case-of-case doesn't fire (either in GHC or Clash), then you should rewrite to:

```haskell
countSuccOverflow e = case e of
  Left a -> let (overflow,a1) = countSuccOverflow a in (False, if overflow then Right countMin else Left a1)
  Right b -> let (overflow,b1) = countSuccOverflow b in (overflow, if overflow then Left countMin else Right b1)
```

Reply:
> I don't know whether GHC does a case-of-case transformation
It does in the simplifier IIRC

Reply:
It only generates one :)